Mathlib/Data/LazyList/Basic.lean

```diff
@@ -159,63 +161,84 @@ instance : Monad LazyList where
   pure := @LazyList.singleton
   bind := @LazyList.bind
 
-theorem append_nil {α} (xs : LazyList α) : xs.append LazyList.nil = xs := by
-  induction xs; rfl
-  simp [LazyList.append, xs_ih]
-  ext; congr
+-- Porting note: Added `Thunk.pure` to definition.
+theorem append_nil {α} (xs : LazyList α) : xs.append (Thunk.pure LazyList.nil) = xs := by
+  induction xs using LazyList.rec; rfl
+  simpa [append]
+  ext; rename_i ih; apply ih
 #align lazy_list.append_nil LazyList.append_nil
 
 theorem append_assoc {α} (xs ys zs : LazyList α) :
-    (xs.append ys).append zs = xs.append (ys.append zs) := by induction xs <;> simp [append, *]
+    (xs.append ys).append zs = xs.append (ys.append zs) := by
+  induction xs using LazyList.rec; rfl
+  simpa [append]
+  ext; rename_i ih; apply ih
 #align lazy_list.append_assoc LazyList.append_assoc
 
+-- Porting note: Rewrote proof of `append_bind`.
 theorem append_bind {α β} (xs : LazyList α) (ys : Thunk (LazyList α)) (f : α → LazyList β) :
-    (@LazyList.append _ xs ys).bind f = (xs.bind f).append ((ys ()).bind f) := by
-  induction xs <;> simp [LazyList.bind, append, *, append_assoc, append, LazyList.bind]
+    (xs.append ys).bind f = (xs.bind f).append (ys.get.bind f) := by
+  match xs with
+  | LazyList.nil => rfl
+  | LazyList.cons x xs =>
+    simp [LazyList.bind, LazyList.append, Thunk.get]
```

Comment:
Please replace these non-terminal simps with the output of `simp?`.
Also elsewhere.
(See https://leanprover-community.github.io/glossary.html#non-terminal-simp)

Reply:
Updated. All non-terminal simps have `only`.
As a side note, `simp?` was slightly broken, so I did some manual work to fix a few simp.